File: sources/Android/android-communications/library/src/test/java/com/polar/sdk/api/model/utils/PolarNightlyRechargeUtilsTest.kt
package com.polar.sdk.api.model.utils

import com.polar.androidcommunications.api.ble.BleLogger
import fi.polar.remote.representation.protobuf.Types.PbDate
import fi.polar.remote.representation.protobuf.Types.PbSystemDateTime
import fi.polar.remote.representation.protobuf.Types.PbTime
import com.polar.androidcommunications.api.ble.model.gatt.client.psftp.BlePsFtpClient
import com.polar.sdk.api.model.sleep.PolarNightlyRechargeData
import com.polar.sdk.impl.utils.PolarNightlyRechargeUtils
import com.polar.sdk.impl.utils.PolarTimeUtils
import fi.polar.remote.representation.protobuf.NightlyRecovery
import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import io.reactivex.rxjava3.core.Single
import org.junit.Test
import protocol.PftpRequest
import java.io.ByteArrayOutputStream
import java.text.SimpleDateFormat
import java.time.LocalDateTime
import java.util.*

class PolarNightlyRechargeUtilsTest {

    @Test
    fun `readNightlyRechargeData() should return nightly recharge data`() {
        // Arrange
        val dateFormatter = SimpleDateFormat("yyyyMMdd", Locale.ENGLISH)
        val mockClient = mockk<BlePsFtpClient>()
        val date = Date()
        val expectedPath = "/U/0/${dateFormatter.format(date)}/NR/NR.BPB"

        val outputStream = ByteArrayOutputStream().apply {
            val proto = NightlyRecovery.PbNightlyRecoveryStatus.newBuilder()
                    .setSleepResultDate(PbDate.newBuilder().setYear(2024).setMonth(12).setDay(5).build())
                    .setCreatedTimestamp(PbSystemDateTime.newBuilder()
                            .setDate(PbDate.newBuilder().setYear(2023).setMonth(12).setDay(5).build())
                            .setTime(PbTime.newBuilder().setHour(10).setMinute(0).setSeconds(0).setMillis(0).build())
                            .setTrusted(true)
                            .build())
                    .setModifiedTimestamp(PbSystemDateTime.newBuilder()
                            .setDate(PbDate.newBuilder().setYear(2023).setMonth(12).setDay(5).build())
                            .setTime(PbTime.newBuilder().setHour(10).setMinute(0).setSeconds(0).setMillis(0).build())
                            .setTrusted(true)
                            .build())
                    .setAnsStatus(5.5f)
                    .setRecoveryIndicator(3)
                    .setRecoveryIndicatorSubLevel(50)
                    .setAnsRate(4)
                    .setScoreRateOBSOLETE(2)
                    .setMeanNightlyRecoveryRRI(800)
                    .setMeanNightlyRecoveryRMSSD(50)
                    .setMeanNightlyRecoveryRespirationInterval(1000)
                    .setMeanBaselineRRI(750)
                    .setSdBaselineRRI(30)
                    .setMeanBaselineRMSSD(45)
                    .setSdBaselineRMSSD(20)
                    .setMeanBaselineRespirationInterval(950)
                    .setSdBaselineRespirationInterval(25)
                    .setSleepTip("Sleep tip 1")
                    .setVitalityTip("Vitality tip 2")
                    .setExerciseTip("Exercise tip 3")
                    .build()
            proto.writeTo(this)
        }

        val createdTimestamp = LocalDateTime.of(2023, 12, 5, 10, 0, 0, 0)
        val calendar = Calendar.getInstance().apply {
            set(2024, Calendar.DECEMBER, 5, 0, 0, 0)
            set(Calendar.MILLISECOND, 0)
        }
        val sleepResultDate = calendar.time

        val expectedResult = PolarNightlyRechargeData(
                createdTimestamp = createdTimestamp,
                modifiedTimestamp = createdTimestamp,
                ansStatus = 5.5f,
                recoveryIndicator = 3,
                recoveryIndicatorSubLevel = 50,
                ansRate = 4,
                scoreRateObsolete = 2,
                meanNightlyRecoveryRRI = 800,
                meanNightlyRecoveryRMSSD = 50,
                meanNightlyRecoveryRespirationInterval = 1000,
                meanBaselineRRI = 750,
                sdBaselineRRI = 30,
                meanBaselineRMSSD = 45,
                sdBaselineRMSSD = 20,
                meanBaselineRespirationInterval = 950,
                sdBaselineRespirationInterval = 25,
                sleepTip = "Sleep tip 1",
                vitalityTip = "Vitality tip 2",
                exerciseTip = "Exercise tip 3",
                sleepResultDate = sleepResultDate
        )

        every { mockClient.request(any()) } returns Single.just(outputStream)

        // Act
        val testObserver = PolarNightlyRechargeUtils.readNightlyRechargeData(mockClient, date).test()

        // Assert
        testObserver.assertComplete()
        testObserver.assertNoErrors()
        testObserver.assertValue(expectedResult)

        verify {
            mockClient.request(
                    PftpRequest.PbPFtpOperation.newBuilder()
                            .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                            .setPath(expectedPath)
                            .build()
                            .toByteArray()
            )
        }
        confirmVerified(mockClient)
    }

    @Test
    fun `readNightlyRechargeData() should return null when an error is thrown`() {
        // Arrange
        val dateFormatter = SimpleDateFormat("yyyyMMdd", Locale.ENGLISH)
        val mockClient = mockk<BlePsFtpClient>()
        val date = Date()
        val expectedPath = "/U/0/${dateFormatter.format(date)}/NR/NR.BPB"
        val expectedError = Throwable("No nightly recharge data found")

        every { mockClient.request(any()) } returns Single.error(expectedError)

        // Act
        val testObserver = PolarNightlyRechargeUtils.readNightlyRechargeData(mockClient, date).test()

        // Assert
        testObserver.assertComplete()
        testObserver.assertNoErrors()
        testObserver.assertNoValues()

        verify {
            mockClient.request(
                    PftpRequest.PbPFtpOperation.newBuilder()
                            .setCommand(PftpRequest.PbPFtpOperation.Command.GET)
                            .setPath(expectedPath)
                            .build()
                            .toByteArray()
            )
        }
        confirmVerified(mockClient)
    }
}

File: sources/iOS/ios-communications/Tests/PolarBleSdkTests/PolarNightlyRechargeUtilsTest.swift
import XCTest
@testable import PolarBleSdk

class PolarNightlyRechargeUtilsTests: XCTestCase {
    
    var mockClient: MockBlePsFtpClient!

    override func setUpWithError() throws {
        mockClient = MockBlePsFtpClient(gattServiceTransmitter: MockPolarGattServiceTransmitter())
    }
    
    override func tearDownWithError() throws {
        mockClient = nil
    }

    func testReadNightlyRechargeData_shouldReturnNightlyRechargeData() async throws {
        // Arrange
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyyMMdd"
        let date = Date()
        let expectedPath = "/U/0/\(dateFormatter.string(from: date))/NR/NR.BPB"

        let proto = Data_PbNightlyRecoveryStatus.with {
            $0.sleepResultDate = PbDate.with { $0.year = 2024; $0.month = 12; $0.day = 5 }
            $0.createdTimestamp = PbSystemDateTime.with {
                $0.date = PbDate.with { $0.year = 2023; $0.month = 12; $0.day = 5 }
                $0.time = PbTime.with { $0.hour = 10; $0.minute = 0; $0.seconds = 0; $0.millis = 0 }
                $0.trusted = true
            }
            $0.modifiedTimestamp = PbSystemDateTime.with {
                $0.date = PbDate.with { $0.year = 2023; $0.month = 12; $0.day = 5 }
                $0.time = PbTime.with { $0.hour = 10; $0.minute = 30; $0.seconds = 0; $0.millis = 0 }
                $0.trusted = true
            }
            $0.ansStatus = 5.5
            $0.recoveryIndicator = 3
            $0.recoveryIndicatorSubLevel = 50
            $0.ansRate = 4
            $0.scoreRateObsolete = 2
            $0.meanNightlyRecoveryRri = 800
            $0.meanNightlyRecoveryRmssd = 50
            $0.meanNightlyRecoveryRespirationInterval = 1000
            $0.meanBaselineRri = 750
            $0.sdBaselineRri = 30
            $0.meanBaselineRmssd = 45
            $0.sdBaselineRmssd = 20
            $0.meanBaselineRespirationInterval = 950
            $0.sdBaselineRespirationInterval = 25
            $0.sleepTip = "Sleep tip 1"
            $0.vitalityTip = "Vitality tip 2"
            $0.exerciseTip = "Exercise tip 3"
        }

        mockClient.requestReturnValue = .success(try proto.serializedData())

        let createdTimestamp = DateComponents(calendar: Calendar.current, year: 2023, month: 12, day: 5, hour: 10, minute: 0).date!
        let modifiedTimestamp = DateComponents(calendar: Calendar.current, year: 2023, month: 12, day: 5, hour: 10, minute: 30).date!
        let sleepResultDate = DateComponents(year: 2024, month: 12, day: 5)

        let expectedResult = PolarNightlyRechargeData(
            createdTimestamp: createdTimestamp,
            modifiedTimestamp: modifiedTimestamp,
            ansStatus: 5.5,
            recoveryIndicator: 3,
            recoveryIndicatorSubLevel: 50,
            ansRate: 4,
            scoreRateObsolete: 2,
            meanNightlyRecoveryRRI: 800,
            meanNightlyRecoveryRMSSD: 50,
            meanNightlyRecoveryRespirationInterval: 1000,
            meanBaselineRRI: 750,
            sdBaselineRRI: 30,
            meanBaselineRMSSD: 45,
            sdBaselineRMSSD: 20,
            meanBaselineRespirationInterval: 950,
            sdBaselineRespirationInterval: 25,
            sleepTip: "Sleep tip 1",
            vitalityTip: "Vitality tip 2",
            exerciseTip: "Exercise tip 3",
            sleepResultDate: sleepResultDate
        )

        // Act
        let testResult = await PolarNightlyRechargeUtils.readNightlyRechargeData(client: mockClient, date: date)

        // Assert
        XCTAssertNotNil(testResult)
        XCTAssertEqual(testResult?.createdTimestamp, expectedResult.createdTimestamp)
        XCTAssertEqual(testResult?.modifiedTimestamp, expectedResult.modifiedTimestamp)
        XCTAssertEqual(testResult?.ansStatus, expectedResult.ansStatus)
        XCTAssertEqual(testResult?.recoveryIndicator, expectedResult.recoveryIndicator)
        XCTAssertEqual(testResult?.recoveryIndicatorSubLevel, expectedResult.recoveryIndicatorSubLevel)
        XCTAssertEqual(testResult?.ansRate, expectedResult.ansRate)
        XCTAssertEqual(testResult?.scoreRateObsolete, expectedResult.scoreRateObsolete)
        XCTAssertEqual(testResult?.meanNightlyRecoveryRRI, expectedResult.meanNightlyRecoveryRRI)
        XCTAssertEqual(testResult?.meanNightlyRecoveryRMSSD, expectedResult.meanNightlyRecoveryRMSSD)
        XCTAssertEqual(testResult?.meanNightlyRecoveryRespirationInterval, expectedResult.meanNightlyRecoveryRespirationInterval)
        XCTAssertEqual(testResult?.meanBaselineRRI, expectedResult.meanBaselineRRI)
        XCTAssertEqual(testResult?.sdBaselineRRI, expectedResult.sdBaselineRRI)
        XCTAssertEqual(testResult?.meanBaselineRMSSD, expectedResult.meanBaselineRMSSD)
        XCTAssertEqual(testResult?.sdBaselineRMSSD, expectedResult.sdBaselineRMSSD)
        XCTAssertEqual(testResult?.meanBaselineRespirationInterval, expectedResult.meanBaselineRespirationInterval)
        XCTAssertEqual(testResult?.sdBaselineRespirationInterval, expectedResult.sdBaselineRespirationInterval)
        XCTAssertEqual(testResult?.sleepTip, expectedResult.sleepTip)
        XCTAssertEqual(testResult?.vitalityTip, expectedResult.vitalityTip)
        XCTAssertEqual(testResult?.exerciseTip, expectedResult.exerciseTip)
        XCTAssertEqual(testResult?.sleepResultDate, expectedResult.sleepResultDate)
        XCTAssertEqual(mockClient.requestCalls.count, 1)

        let actualPath = try Protocol_PbPFtpOperation(serializedBytes: mockClient.requestCalls[0]).path
        XCTAssertEqual(actualPath, expectedPath)
    }
    
    func testReadNightlyRechargeFromDayDirectory_FileNotFound() async {
        // Arrange
        mockClient.requestReturnValue = .failure(NSError(domain: "File not found", code: 103, userInfo: nil))

        // Act — errors are swallowed; method returns nil
        let result = await PolarNightlyRechargeUtils.readNightlyRechargeData(client: mockClient, date: Date())

        // Assert
        XCTAssertNil(result, "Expected nil when nightly recovery file is not found")
    }
}
